pkg/mutation/schema/schema.go

```diff
@@ -0,0 +1,398 @@
+		// We removed all changes made by the previous mutator and
+		// are re-adding a mutator that was already present. Because
+		// this mutator was already present and we have a lock on the
+		// db, this should never fail. If it does we are in an unknown
+		// state and should panic so we can recover by bootstrapping
+		// and raise the visibility of the issue.
+		log.Error(err, "could not upsert previously existing mutator into schema, this is not recoverable")
+		panic(err)
+	}
+}
+
+// Remove removes the mutator with the given id from the
+// db.
+func (db *DB) Remove(id types.ID) {
+	db.mutex.Lock()
+	defer db.mutex.Unlock()
+	db.remove(id)
+	delete(db.mutators, id)
+}
+
+func (db *DB) remove(id types.ID) {
+	mutator, ok := db.mutators[id]
+	if !ok {
+		// no mutator found, nothing to do
+		return
+	}
+
+	iterator := db.bindingIterator.over(mutator.SchemaBindings())
+	for gvk := iterator(); gvk != nil; gvk = iterator() {
+		s, ok := db.schemas[*gvk]
+		if !ok {
+			log.Error(nil, "mutator associated with missing schema", "mutator", id, "schema", *gvk)
+			panic(fmt.Sprintf("mutator %v associated with missing schema %v", id, *gvk))
+		}
+		s.remove(mutator.Path().Nodes)
+		if s.root == nil {
+			delete(db.schemas, *gvk)
+		}
+	}
+}
+
+type scheme struct {
+	gvk  schema.GroupVersionKind
+	root *node
+}
+
+func (s *scheme) add(ref []parser.Node) error {
+	if s.root == nil {
+		s.root = &node{}
+	}
+	return s.root.add(ref)
+}
+
+func (s *scheme) remove(ref []parser.Node) {
+	if s.root == nil {
+		return
+	}
+	s.root.remove(ref)
+	if s.root.referenceCount == 0 {
+		s.root = nil
+	}
+}
+
+type node struct {
+	referenceCount uint
+	nodeType       parser.NodeType
+
+	// list-type nodes have a key field and only one child
+	keyField *string
+	child    *node
+
+	// object-type nodes only have children
+	children map[string]*node
+}
+
+// backup creates a shallow copy of the node we can restore in case of error
+func (n *node) backup() *node {
+	return &node{
+		referenceCount: n.referenceCount,
+		nodeType:       n.nodeType,
+		keyField:       n.keyField,
+	}
+}
+
+func (n *node) restore(backup *node) {
+	n.referenceCount = backup.referenceCount
+	n.nodeType = backup.nodeType
+	n.keyField = backup.keyField
+}
+
+func (n *node) add(ref []parser.Node) error {
+	backup := n.backup()
+	n.referenceCount++
+
+	// we should stop recursing when len(ref) == 1, but
+	// this guards against infinite recursion in the case
+	// where there is a bug in the switch code below
+	if len(ref) == 0 {
+		return nil
+	}
+
+	err := func() error {
+		current := ref[0]
+		switch t := current.Type(); t {
+		case parser.ObjectNode:
+			if n.nodeType != "" && n.nodeType != parser.ObjectNode {
+				return fmt.Errorf("node type conflict: %v vs %v", n.nodeType, parser.ObjectNode)
+			}
+			n.nodeType = parser.ObjectNode
+			obj := current.(*parser.Object)
+			if n.children == nil {
+				n.children = make(map[string]*node)
+			}
+			if len(ref) == 1 {
```

Comment:
This may look a bit obscure. Worth adding a comment explaining we are not adding the leaves and why?

Reply:
+1 good point, adding a comment